torch/csrc/jit/passes/create_autodiff_subgraphs.cpp

```diff
@@ -1,257 +1,293 @@
 #include "torch/csrc/jit/passes/create_autodiff_subgraphs.h"
 
-#include "torch/csrc/jit/ir.h"
-#include "torch/csrc/jit/autodiff.h"
 #include "torch/csrc/jit/assertions.h"
-#include "torch/csrc/jit/dynamic_dag.h"
-
-#include <cstddef>
-#include <limits>
-
-namespace torch { namespace jit {
+#include "torch/csrc/jit/autodiff.h"
+#include "torch/csrc/jit/ir.h"
+#include "torch/csrc/jit/passes/common_subexpression_elimination.h"
 
-struct Graph;
+namespace torch {
+namespace jit {
 
 namespace {
 
-// Move nodes that exist in graph g into a 'group_node_kind' node.
-// All inputs shared by the nodes become inputs to the new node.
-// Outputs from 'nodes' are redirected to outputs of the new node,
-// and the original nodes are removed.
-// prereq: it is topologically valid to place the new node
-// right before nodes[0] (i.e. it will not create cycles and all uses of
-// new node will be after this position).
-// prereq: nodes are in topological order
-Node* mergeNodes(Block * block, Symbol group_node_kind, ArrayRef<Node*> nodes) {
-  JIT_ASSERT(nodes.size() > 0);
-  std::unordered_map<Value*, Value*> value_map;
-  Graph * graph = block->owningGraph();
-
-  auto new_graph = std::make_shared<Graph>();
-  Node * group_node = graph->create(group_node_kind, 0);
-  group_node->g_(attr::Subgraph, new_graph);
-
-  auto getOrCreateInput = [&](Value * v) {
-    if(value_map.count(v) > 0) {
-      return value_map[v];
-    }
-    if (auto value = toIValue(v)) {
-      Value * nv = new_graph->insertConstant(*value);
-      value_map[v] = nv;
-      return nv;
+class SubgraphSlicer {
+ public:
+  SubgraphSlicer(Block* block, size_t minSubgraphSize)
+      : block_(block), minSubgraphSize_(minSubgraphSize) {}
+
+  void run(std::vector<Node*>& diffGraphs) {
+    bool any_changed = true;
+    while (any_changed) {
```

Comment:
The previous approach scans the graph once from `rbegin` to `rend`: why does this need to scan through multiple times?

Reply:
This is because of an implementation detail of `moveBeforeTopologicallyValid`. We may move nodes after the insertion point, so it's possible we need to do multiple passes (similar to the graph fuser).

Reply:
hmm `a.moveBeforeTopologicallyValid(b)` should only move the nodes between `a` and `b`. Because we're scanning for mergeable groups in reverse topological order, after we're done scanning `b` we should be convinced that there are no more merges that can include `b`, but I'm probably missing something here. Do you have an example for how we might need to do multiple passes here?

Reply:
`a.moveBeforeTopo(b)` will only move nodes between `a` and `b`, but some of those nodes could end up after the insertion point.

For example
```
c = f(a, b)
d = f(c)
e = f(d) <-- we're here in the loop
```
`c.moveBeforeTopologicallyValid(e)` will do:
```
c = f(a, b) 
e = f(d)   <- still here in the loop
d = f(c)   <- this was moved on the other side
```